src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -159,189 +170,260 @@ internal static Enumeration DetermineFileEnumerationFromGlob(string includeGlobP
+            else if (segment.First() == '\'')
             {
-                throw new ArgumentException("A key segment cannot start with a single-quote character`.");
+                throw new ArgumentException($"A {segmentType} segment cannot start with a single-quote character`.");
             }
-            else if (keySegment.First() == '`')
+            else if (segment.First() == '`')
             {
-                throw new ArgumentException("A key segment cannot start with a backtick character`.");
+                throw new ArgumentException($"A {segmentType} segment cannot start with a backtick character`.");
             }
         }
 
-        public static Fingerprint EvaluateKeyToFingerprint(
+        private static void LogSegment(
             AgentTaskPluginExecutionContext context,
-            string filePathRoot,
-            IEnumerable<string> keySegments)
+            IEnumerable<string> segments,
+            string segment,
+            KeySegmentType type,
+            Object details)
         {
-            // Quickly validate all segments
-            foreach (string keySegment in keySegments)
+            Func<string, int, string> FormatForDisplay = (value, displayLength) =>
             {
-                CheckKeySegment(keySegment);
-            }
+                if (value.Length > displayLength)
+                {
+                    value = value.Substring(0, displayLength - 3) + "...";
+                }
 
-            string defaultWorkingDirectory = context.Variables.GetValueOrDefault(
-                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
-                )?.Value;
+                return value.PadRight(displayLength);
+            };
 
-            var resolvedSegments = new List<string>();
-            var exceptions = new List<Exception>();
+            string formattedSegment = FormatForDisplay(segment, Math.Min(segments.Select(s => s.Length).Max(), 50));
 
-            Action<string, KeySegmentType, Object> LogKeySegment = (segment, type, details) =>
+            if (type == KeySegmentType.String)
             {
-                Func<string,int,string> FormatForDisplay = (value, displayLength) =>
+                context.Output($" - {formattedSegment} [string]");
+            }
+            else if (type == KeySegmentType.Directory)
+            {
+                context.Output($" - {formattedSegment} [directory]");
+            }
+            else if (type == KeySegmentType.DirectoryPattern)
+            {
+                var directories = (details as string[]) ?? new string[0];
+                context.Output($" - {formattedSegment} [directory pattern; matches: {directories.Length}]");
+                if (directories.Any())
                 {
-                    if (value.Length > displayLength)
+                    int filePathDisplayLength = Math.Min(directories.Select(d => d.Length).Max(), 70);
```

Comment:
Very similar to FilePattern - can we create a common function to avoid duplication?